deployment_logger.go

```diff
@@ -0,0 +1,248 @@
+		//logger:
+		// for now we can hardcode this
+		maxLogFiles:    5,
+		loggingEnabled: false,
+	}
+}
+
+func (dlm DeploymentLogManager) WriteLog(log []byte) error {
+	if dlm.logger == nil {
+		return ErrLoggerNotInitialized
+	}
+	_, err := dlm.logger.Write(log)
+	return err
+}
+
+func (dlm *DeploymentLogManager) Enable() error {
+	if dlm.loggingEnabled {
+		// TODO: maybe in fact we should return error
+		return nil
+	}
+
+	// check if we have valid deploymend id
+	if dlm.deploymentID == "" {
+		return ErrNoDeploymentId
+	}
+
+	// check if we should write to last file or create new one
+	depIdFile, err := dlm.GetDeploymentIDFromLogFile()
+	if err != nil {
+		// we are creating first log file
+		if err == ErrNoLogFiles {
+			depIdFile = dlm.deploymentID
+		} else {
+			// TODO: what if file exists but is broken?
+			return err
+		}
+	}
+
+	// we have new deployment so we need to rotate files
+	if depIdFile != dlm.deploymentID {
+		dlm.Rotate()
+	}
+
+	// instanciate logger
+	dlm.logger = NewFileLogger(filepath.Join(dlm.logLocation, dlm.logFileName))
+	if dlm.logger == nil {
+		return ErrLoggerNotInitialized
+
+	}
+
+	dlm.loggingEnabled = true
+	return nil
+}
+
+func (dlm *DeploymentLogManager) Disable() error {
+	if !dlm.loggingEnabled {
+		// TODO: maybe in fact we should return error
+		return nil
+	}
+
+	if err := dlm.logger.Deinit(); err != nil {
+		return err
+	}
+
+	dlm.loggingEnabled = false
+	return nil
+}
+
+func (dlm DeploymentLogManager) getSortedLogFiles() ([]string, error) {
+
+	// list all the log files in log directory
+	fileExtension := filepath.Ext(dlm.logFileName)
+	logFiles, err :=
+		filepath.Glob(filepath.Join(dlm.logLocation, dlm.logFileName[0:len(dlm.logFileName)-len(fileExtension)]) + "*")
+	if err != nil {
+		return nil, err
+	}
+
+	sort.Sort(sort.Reverse(sort.StringSlice(logFiles)))
+	return logFiles, nil
+}
+
+func (dlm DeploymentLogManager) rotateLogFileName(name string) string {
+	fileExtension := filepath.Ext(name)
+	seq, err := strconv.Atoi(fileExtension[1:])
+	if err == nil {
+		// IDEA: this will allow handling 9999 log files correctly
+		// for more we need to change implementation of getSortedLogFiles()
+		return fmt.Sprintf("%s.%04d", name[0:len(name)-len(fileExtension)], seq+1)
+	}
+	return name + ".0001"
+}
+
+// we are storing old log files adding `.seqnum` suffix to log file name
+func (dlm DeploymentLogManager) Rotate() {
+	logFiles, err := dlm.getSortedLogFiles()
+	if err != nil {
+		// can not rotate
+		return
+	}
+
+	// check if we need to delete oldest file
+	for len(logFiles) > dlm.maxLogFiles {
+		os.Remove(logFiles[0])
+		logFiles = append(logFiles[1:])
+	}
+
+	// rename log files; only those not removed
+	for i := range logFiles {
+		os.Rename(logFiles[i], dlm.rotateLogFileName(logFiles[i]))
+	}
+}
+
+// we need to check what was last deployment ID logged in case of reboot or power-loss
+func (dlm DeploymentLogManager) GetDeploymentIDFromLogFile() (string, error) {
```

Comment:
IMO this is not part of deployment logging. Update state information should be preserved elsewhere, perhaps we need to add state context passed between update states.

What I'm missing is a way of obtaining the logs for a particular deployment ID. This is something we need in UpdateError state where we should load logs of the deployment that failed..